sdk/java/libjfs/bridge.go

```diff
@@ -0,0 +1,334 @@
+
+	CommonLabels map[string]string
+}
+
+// Bridge pushes metrics to the configured Graphite server.
+type Bridge struct {
+	useTags  bool
+	url      string
+	prefix   string
+	interval time.Duration
+	timeout  time.Duration
+
+	errorHandling HandlerErrorHandling
+	logger        Logger
+
+	g            prometheus.Gatherer
+	commonLabels map[string]string
+}
+
+// Logger is the minimal interface Bridge needs for logging. Note that
+// log.Logger from the standard library implements this interface, and it is
+// easy to implement by custom loggers, if they don't do so already anyway.
+type Logger interface {
+	Println(v ...interface{})
+}
+
+// NewBridge returns a pointer to a new Bridge struct.
+func NewBridge(c *Config) (*Bridge, error) {
+	b := &Bridge{}
+
+	b.useTags = c.UseTags
+
+	if c.URL == "" {
+		return nil, errors.New("missing URL")
+	}
+	b.url = c.URL
+
+	if c.Gatherer == nil {
+		b.g = prometheus.DefaultGatherer
+	} else {
+		b.g = c.Gatherer
+	}
+
+	if c.Logger != nil {
+		b.logger = c.Logger
+	}
+
+	if c.Prefix != "" {
+		b.prefix = c.Prefix
+	}
+
+	var z time.Duration
+	if c.Interval == z {
+		b.interval = defaultInterval
+	} else {
+		b.interval = c.Interval
+	}
+
+	if c.Timeout == z {
+		b.timeout = defaultInterval
+	} else {
+		b.timeout = c.Timeout
+	}
+
+	b.errorHandling = c.ErrorHandling
+
+	b.commonLabels = c.CommonLabels
+	return b, nil
+}
+
+// Run starts the event loop that pushes Prometheus metrics to Graphite at the
+// configured interval.
+func (b *Bridge) Run(ctx context.Context) {
+	ticker := time.NewTicker(b.interval)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ticker.C:
+			if err := b.Push(); err != nil && b.logger != nil {
+				b.logger.Println("error pushing to Graphite:", err)
+			}
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+// Push pushes Prometheus metrics to the configured Graphite server.
+func (b *Bridge) Push() error {
+	mfs, err := b.g.Gather()
+	if b.commonLabels != nil {
+		for _, mf := range mfs {
+			for _, metric := range mf.Metric {
+				for k, v := range b.commonLabels {
+					metric.Label = append(metric.Label, &dto.LabelPair{
+						Name:  proto.String(k),
+						Value: proto.String(v),
+					})
+				}
+			}
+		}
+	}
+	if err != nil || len(mfs) == 0 {
+		switch b.errorHandling {
+		case AbortOnError:
+			return err
+		case ContinueOnError:
+			if b.logger != nil {
+				b.logger.Println("continue on error:", err)
+			}
+		default:
+			panic("unrecognized error handling value")
+		}
+	}
+
+	conn, err := net.DialTimeout("tcp", b.url, b.timeout)
+	if err != nil {
+		return err
+	}
+	defer conn.Close()
+
+	return writeMetrics(conn, mfs, b.useTags, b.prefix, model.Now())
+}
+
+func writeMetrics(w io.Writer, mfs []*dto.MetricFamily, useTags bool, prefix string, now model.Time) error {
+	vec, err := expfmt.ExtractSamples(&expfmt.DecodeOptions{
+		Timestamp: now,
+	}, mfs...)
+	if err != nil {
+		return err
+	}
+
+	buf := bufio.NewWriter(w)
+	for _, s := range vec {
+		if prefix != "" {
```

Comment:
The original logic here is to still write a point when the prefix is empty. I think this is a bug, so I modified it here. When prefix is empty, dots should not be written
```suggestion
		if prefix != "" {
```

Reply:
This is indeed a bug, upstream has been merged.https://github.com/prometheus/client_golang/pull/1005